### PR TITLE
APPS-752 dxApi utility to clone file, APPS-754 dxApi utility to clone applet. 

### PR DESCRIPTION
**Pull request by r-i-v-a**

### CODE REVIEW

File: api/src/main/scala/dx/api/DxApi.scala [hunk 1]
```diff
@@ -954,6 +954,28 @@ case class DxApi(version: String = "1.0.0", dxEnv: DXEnvironment = DxApi.default
     }
   }
 
+  // Copy object to destination project, if it isn't already there.
+  def cloneObject(id: String,
```
Comment (r-i-v-a) [inline api/src/main/scala/dx/api/DxApi.scala:958]: Question -- I eventually want the ability to clone both files and applets to the current project, if not already there (see https://jira.internal.dnanexus.com/browse/APPS-752 , https://jira.internal.dnanexus.com/browse/APPS-754), it seems like projectClone can take objects of different types, so could it be done this way? @jdidion

File: api/src/main/scala/dx/api/DxApi.scala [hunk 2]
```diff
@@ -954,6 +954,28 @@ case class DxApi(version: String = "1.0.0", dxEnv: DXEnvironment = DxApi.default
...
+      )
+    } else {
+      logger.trace(s"Cloning object ${id} from ${sourceProject.id} to ${destProject.id}")
+      val request = Map("objects" -> JsArray(JsString(id)),
+                        "project" -> JsString(destProject.id),
+                        "destination" -> JsString(destFolder))
+      projectClone(sourceProject.id, request)
+
+      // TODO: Is this all that is necessary for cloning only file, not record?
+      // TODO: Could it work this way for both files and applets?
```
Comment (r-i-v-a) [inline api/src/main/scala/dx/api/DxApi.scala:975]: Also, I looked at cloneAsset as an example, but want to exclude the parts that are only needed when cloning a record -- is it correct? @jdidion

**Comment by jdidion** [comment]
Please update the function name to `cloneDataObject`. "Data object" is platform terminology for any object that exists within the context of a project (e.g. file, record, applet, workflow).

**Comment by r-i-v-a** [comment]
> Please update the function name to `cloneDataObject`. "Data object" is platform terminology for any object that exists within the context of a project (e.g. file, record, applet, workflow).

Done.

**Review by jdidion — CHANGES_REQUESTED** [review]

File: api/src/main/scala/dx/api/DxApi.scala [hunk 3]
```diff
@@ -954,6 +954,25 @@ case class DxApi(version: String = "1.0.0", dxEnv: DXEnvironment = DxApi.default
...
+                  destProject: DxProject,
+                  destFolder: String = "/"): Unit = {
+    if (sourceProject.id == destProject.id) {
+      logger.trace(
+          s"""The source and destination projects are the same (${sourceProject.id}),
+             |no need to clone object ${id}""".stripMargin.replaceAll("\n", " ")
+      )
+    } else {
+      logger.trace(s"Cloning object ${id} from ${sourceProject.id} to ${destProject.id}")
+      val request = Map("objects" -> JsArray(JsString(id)),
```
[the remark this answers is not in the document]
Comment (r-i-v-a) [reply api/src/main/scala/dx/api/DxApi.scala:969]: https://documentation.dnanexus.com/developer/api/data-containers/cloning#api-method-class-xxxx-clone

[the remark this answers is not in the document]
Comment (r-i-v-a) [reply api/src/main/scala/dx/api/DxApi.scala:969]: Setting parents -> true in both functions.

File: api/src/main/scala/dx/api/DxApi.scala [hunk 4]
```diff
@@ -919,7 +919,8 @@ case class DxApi(version: String = "1.0.0", dxEnv: DXEnvironment = DxApi.default
       logger.trace(s"Cloning asset ${assetName} from ${sourceProject.id} to ${destProject.id}")
       val request = Map("objects" -> JsArray(JsString(assetRecord.id)),
                         "project" -> JsString(destProject.id),
-                        "destination" -> JsString(destFolder))
+                        "destination" -> JsString(destFolder),
+                        "parents" -> JsBoolean(true))
```
Comment (jdidion) [inline api/src/main/scala/dx/api/DxApi.scala:923]: Minor comment - you can use `JsTrue` rather than `JsBoolean(true)`

**Review by jdidion — APPROVED** [review]

**Pull request merged**